/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

syntax = "proto2";

package rpcenv;

message Action {
  optional int32 action = 1;
  optional ArrayNest nest_action = 2;
  optional int32 save_env = 3;
  optional int32 env_data_label = 4;
  optional NDArray env_data = 5;
}

message NDArray {
  optional int32 dtype = 1;
  repeated int64 shape = 2 [packed = true];
  optional bytes data = 3;
};

message ArrayNest {
  optional NDArray array = 1;
  repeated ArrayNest vector = 2;
  map<string, ArrayNest> map = 3;
};

message Step {
  optional ArrayNest observation = 1;
  optional float reward = 2;
  optional bool done = 3;
  optional bool reset_done = 4;
  optional int32 episode_step = 5;
  optional float episode_return = 6;
  optional int32 episode_server = 7;
  optional int32 episode_count = 8;
  optional NDArray env_data = 9;
  optional ArrayNest last_observation = 10;
}

service RPCEnvServer {
  rpc StreamingEnv(stream Action) returns (stream Step) {}
}
